File: NeetCode/Arrays_and_Hashing/products_of_array_except_self/products_of_array_except_self.py
from typing import Optional, List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        left = [
            1,
        ] * len(nums)
        right = left[:]
        left[0] = 1
        right[-1] = 1

        for index in range(1, len(nums)):
            left[index] = left[index - 1] * nums[index - 1]
        for index in range(len(nums) - 2, -1, -1):
            right[index] = right[index + 1] * nums[index + 1]

        return [left[index] * right[index] for index in range(len(nums))]


if __name__ == "__main__":
    sol = Solution()

    nums = [1, 2, 4, 6]
    print(sol.productExceptSelf(nums))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/products_of_array_except_self/products_of_array_except_self.cpp
// products_of_array_except_self.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> productExceptSelf(vector<int> &nums)
    {
        int n = nums.size();
        int r = 1;
        vector<int> result(n, 1);

        for (int j = 1; j < n; j++)
        {
            result[j] = result[j - 1] * nums[j - 1];
        }
        for (int i = n - 1; i >= 0; i--)
        {
            result[i] = result[i] * r;
            r *= nums[i];
        }

        return result;
    }
};

int main()
{
    Solution sol;

    vector<int> test = {1, 2, 3, 4};

    vector<int> result = sol.productExceptSelf(test);

    for (int n : result)
    {
        cout << n << " ";
    }
    cout << endl;

    return 0;
}
